### PR TITLE
Feat/ruff formatter

### LINKED ISSUE #203: automated formatter for python code

### PR DESCRIPTION
**Pull request by landonshumway-ia**
We have determined to add Ruff as our formatter for this project. Ruff can also be used as a linter, and so we are replacing pylint to reduce the number of dependencies we have to work with to maintain this project. This introduces the Ruff dependency, and replaces the pylint commands with the Ruff check command. See [Ruff tutorial][0]

We have selected a reasonable set of linting/formatting rules to start with, though we expect to adjust the rules of this tool as we continue to use it. This PR introduces all of the formatting/linting recommended changes so that every python file in the project is using a consistent formatting according to this beginning rule set. Because this PR touches every python file in the project, in the interest of avoiding significant merge conflicts, it is requested that this PR not be held up over decisions of specific ruff rules to apply, as those can be added in subsequent PRs. The intent of this PR is to introduce the tooling with consistent formatting applied.

**It is important to note that this PR is intended to be a pure refactor with NO runtime code implementation changes.**

[0]: https://docs.astral.sh/ruff/linter/

### Requirements List
- Added new Ruff dependency to `compact-connect/requirements-dev.txt`, which can be installed locally to run the 
linter/formatter

### Description List
- Replace pylintrc with ruff.toml for linter/formatter configuration rules.
- Ran automatic linter fixes, which includes the following:
   - sorting import in alphabetical order
   - removing redundant 'r' arg when opening files
   - replacing deprecated type imports with recommended collections module
- Ran ruff formatter using recommended default options
- Replaced pylint comments with ruff comments where applicable

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- For API configuration changes: CDK tests added/updated in `backend/compact-connect/tests/unit/test_api.py`
- Code review

Closes #203

### CODE REVIEW

File: .github/workflows/check-python.yml
Comment (jlkravitz) [inline .github/workflows/check-python.yml:31]: unless you intentionally want to _ignore_ .py files that are not part of the git repository, you should be able to run `ruff check` to recursively lint all python files.

(same comment for `format_python.sh`)

**Review by jlkravitz — COMMENTED** [review]
one question/comment but otherwise looks great!

File: .github/workflows/check-python.yml
Comment (landonshumway-ia) [reply .github/workflows/check-python.yml:31]: I'm good either way. I can see the benefit of not checking files that are not added to the repo, as sometimes I have scratch files that I don't want the shell script to flag me on. But if a file is checked into the git repository, obviously we will want to check it. My personal vote is to keep this as it is, but I'm curious if @jusdino has a different opinion.

Comment (jusdino) [reply .github/workflows/check-python.yml:31]: I don't feel strongly either, but generally agree with @landonshumway-ia - the check is intended to be specific to files that are checked in. Of course, given that this is run only against the GitHub repo, freshly checked out by automation, I think the distinction is moot anyway.

Comment (jlkravitz) [reply .github/workflows/check-python.yml:31]: Sounds good. That's fine with me!

**Review by jlkravitz — DISMISSED** [review]
@isabeleliassen good to merge!

**Review by isabeleliassen — DISMISSED** [review]

**Comment by landonshumway-ia** [comment]
@jlkravitz There was a merge conflict between this branch and #248. I've resolved the conflicts and re-ran the formatter/linter, but it looks like it will need another approval from you and @isabeleliassen before it can be merged.

**Review by jlkravitz — APPROVED** [review]

**Review by isabeleliassen — APPROVED** [review]

**Pull request merged**
